### PR TITLE
Share logic between setup flow elevated process and Dev Home process

### LINKED ISSUE #622: Refactor and share code between elevated and dev home processes
## Changes:
- Share common code between elevated and Dev Home code
- Ensure task names and methods signatures are consistent

## Related:
- https://github.com/microsoft/devhome/issues/1251

### PR DESCRIPTION
**Pull request by AmelBawa-msft**
## Closing PR Note
For now I am closing this PR as we're updating how the Dev Home process communicates to the elevated process. Changes in this PR and the comments here will be used as a reference for any future work related to sharing code between elevated and non-elevated process.

## Summary of the pull request
This PR suggests an alternative approach for simplifying the implementation of setup flow operations such as DSC configuration, app installation and dev drive creation. The change aims to enable the setup flow elevated process to benefit from the existing Dev Home code implementation:
- Single code implementation for Dev Home and the Elevated process. New features are added a single time only.
- Types returned by the Elevated process are compatible with the types in Dev Home process, hence eliminating the current need for adapter classes.
- `Execute` and `ExecuteAsAdmin` can now share code as the former will have an input factory instance that lives in-proc, whereas the former will have an input factory instance (of the same static type) that lives out-of-proc (elevated proc).
```cs
// Result interface from Dev Home or the elevated process
private IInstallPackageResult_result;

public IAsyncOperation<TaskFinishedState> Execute(ITaskOperatorFactory operatorFactory)
{
    // Today {operatorFactory} lives in proc, but moving it out of proc is forward-compatible with the current signature.
    // https://github.com/microsoft/devhome/issues/1251
    var operator = operatorFactory.CreateInstallPackageOperator();
    _result = operator.InstallPackageAsync(...);
}

public IAsyncOperation<TaskFinishedState> ExecuteAsAdmin(ITaskOperatorFactory elevatedOperatorFactory)
{
    // {elevatedOperatorFactory} is a proxy for an out-of-proc COM object that lives in the elevated process
    var operator = operatorFactory.CreateInstallPackageOperator();
    _result = operator.InstallPackageAsync(...);
}
```

### Changes
- Removed `DevHome.SetupFlow.ElevatedComponent` and `DevHome.SetupFlow.ElevatedComponent.Projection` csproj which contained a CsWinRTComponent for contracts and implementation which included the duplicate logic addressed in this PR.
- Created `DevHome.SetupFlow.Contract` vcxproj containing an IDL for contract interfaces producing only a WINMD.
- Implemented the CsWinRT projected interfaces in the `DevHome.SetupFlow`
- From the elevated process project, add a reference to the `DevHome.SetupFlow` csproj

## PR checklist
- [ ] Closes #622
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: DevHome.sln [hunk 1]
```diff
@@ -58,20 +58,12 @@ EndProject
...
-	ProjectSection(ProjectDependencies) = postProject
-		{A7E5FD7B-B41A-4CAE-A45A-E686DFA8ACF1} = {A7E5FD7B-B41A-4CAE-A45A-E686DFA8ACF1}
-	EndProjectSection
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "CoreWidgetProvider", "CoreWidgetProvider", "{32C0052F-10AF-48C0-A7A3-B0A1793397FD}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CoreWidgetProvider", "CoreWidgetProvider\CoreWidgetProvider.csproj", "{0D879E08-99AA-4019-9D04-DEA9F7C7BFC1}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CoreWidgetProvider", "CoreWidgetProvider\CoreWidgetProvider.csproj", "{0D879E08-99AA-4019-9D04-DEA9F7C7BFC1}"
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "DevHome.SetupFlow.Contract", "tools\SetupFlow\DevHome.SetupFlow.Contract\DevHome.SetupFlow.Contract.vcxproj", "{493E025E-AA0A-44A2-A2A5-6DA88329194C}"
```
Comment (florelis) [inline DevHome.sln:66]: Nit: I'm not a fan of the name `DevHome.SetupFlow.Contract`. `Contract` seems too generic for how specific it is; maybe something like `TaskResultContract`?

File: src/DevHome.csproj [hunk 2]
```diff
@@ -46,6 +46,7 @@
         <ProjectReference Include="..\CoreWidgetProvider\CoreWidgetProvider.csproj" />
         <ProjectReference Include="..\settings\DevHome.Settings\DevHome.Settings.csproj" />
         <ProjectReference Include="..\tools\Dashboard\DevHome.Dashboard\DevHome.Dashboard.csproj" />
+        <ProjectReference Include="..\tools\SetupFlow\DevHome.SetupFlow.ElevatedServer\DevHome.SetupFlow.ElevatedServer.csproj" />
```
Comment (florelis) [inline src/DevHome.csproj:49]: Why is this new reference needed?

File: tools/SetupFlow/DevHome.SetupFlow.ElevatedServer/DevHome.SetupFlow.ElevatedServer.csproj [hunk 3]
```diff
@@ -20,16 +20,6 @@
...
-  </ItemGroup>
-
-  <!-- The server .exe requires the .winmd to be located next to it. The project reference 
-        is not enough to have it automatically placed there. Manually include it in the outputs. -->
-  <ItemGroup>
-    <Content Include="..\DevHome.SetupFlow.ElevatedComponent\bin\$(Platform)\$(Configuration)\$(TargetFramework)\DevHome.SetupFlow.ElevatedComponent.winmd">
-      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
-      <CopyToPublishDirectory>Always</CopyToPublishDirectory>
-    </Content>
+    <ProjectReference Include="..\DevHome.SetupFlow\DevHome.SetupFlow.csproj" />
```
Comment (florelis) [inline tools/SetupFlow/DevHome.SetupFlow.ElevatedServer/DevHome.SetupFlow.ElevatedServer.csproj:23]: I'm not a fan of having the elevated process depend on the full SetupFlow project. That makes the elevated server depend on the whole setup flow, which depends on the elevated server. It's not an issue for the compiler, but I'm still not comfortable with that.
I would put the shared parts in the Common project, and keep there only the minimum we need. 

But if we're going to keep the dependency on the SetupFlow, I think we can do away with the Common project altogether

File: DevHome.sln (hunk 1 above)
Comment (asklar) [reply DevHome.sln:66]: shouldn't this be part of the SDK? why a separate project?
cc @EricJohnson327

File: tools/SetupFlow/DevHome.SetupFlow/TaskOperator/ConfigurationOperator.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    public IAsyncAction OpenConfigurationSetAsync(StorageFile file)
+    {
+        return Task.Run(async () =>
+        {
+            Log.Logger?.ReportInfo(Log.Component.Configuration, $"Opening configuration set from file: {file.Path}");
+
+            try
+            {
+                var modulesPath = Path.Combine(AppContext.BaseDirectory, @"runtimes\win\lib\net6.0\Modules");
```
Comment (asklar) [inline tools/SetupFlow/DevHome.SetupFlow/TaskOperator/ConfigurationOperator.cs:34]: what happens when we migrate to .net 8?

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        interface IInstallPackageResult
+            requires IOperationResult
+        {
+            // Gets the status of the install call
+            Int32 Status { get; };
+
+            // Gets the installer error code
+            UInt32 InstallerErrorCode { get; };
+
+            // Gets the error code of the overall install operation
```
[the remark this answers is not in the document]
Comment (asklar) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:28]: yeah reading the interface I have no idea if I'm supposed to return an hresult, an ntstatus, something else

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            Int32 HResult { get; };
+        };
+
+        interface IConfigurationUnitResult
+        {
+            // Gets the name of the configuration unit
+            String UnitName { get; };
+
+            // Gets the intent of the configuration unit (e.g. Assert, Inform, Apply)
+            String Intent { get; };
```
Comment (asklar) [inline tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:45]: why is this a string? why not an enum? is this shown to the user? if so, how does this get localized?

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        };
+
+        interface ICreateDevDriveResult
+            requires IOperationResult
+        {
+            // Gets the HRESULT value of the operation
+            Int32 HResult { get; };
+        };
+
+        interface IConfigurationUnitResult
```
Comment (asklar) [inline tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:39]: what is a configuration unit? not clear from this definition. does it refer to a configuration being N configuration units?

Comment (asklar) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:39]: prefer instead: N Configuration --> 1 ConfigurationSet

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            Windows.Foundation.Collections.IVector<IConfigurationUnitResult> UnitResults { get; };
+        };
+
+        interface IInstallPackageOperator
+        {
+            // Install package
+            Windows.Foundation.IAsyncOperation<IInstallPackageResult> InstallPackageAsync(String packageId, String catalogName);
+        };
+
+        interface IDevDriveStorageOperator
```
Comment (asklar) [inline tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:67]: all these aren't operators, they're operations :) prefer standard naming

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl (hunk 7 above)
Comment (florelis) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:39]: Configuration Unit/Configuration Set is what's used in the configuration API: https://github.com/microsoft/winget-cli/blob/master/src/Microsoft.Management.Configuration/Microsoft.Management.Configuration.idl

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        // This operator exists just for testing
+        interface ITestOperator
+        {
+            // Write to standard output
+            void WriteToStdOut(String msg);
+        };
+
+        interface ITaskOperatorFactory
```
Comment (asklar) [inline tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:89]: SetupOperationFactory

Comment (asklar) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:89]: does this even need to be an interface? these are all essentially static methods, can be put on a runtimeclass

File: DevHome.sln (hunk 1 above)
Comment (florelis) [reply DevHome.sln:66]: I don't think it needs to be part of the SDK unless we want to allow extensions that extend the setup flow. And that would be an item bigger than this change.

File: tools/SetupFlow/DevHome.SetupFlow/Models/InstallPackageTask.cs [hunk 10]
```diff
@@ -93,78 +92,72 @@ public ActionCenterMessages GetRebootMessage()
...
-                _installResultStatus = e.Status;
-                _extendedErrorCode = e.ExtendedErrorCode;
-                _installerErrorCode = e.InstallerErrorCode;
-                ReportAppInstallFailedEvent();
-                Log.Logger?.ReportError(Log.Component.AppManagement, $"Failed to install package with status {e.Status} and installer error code {e.InstallerErrorCode}");
-                return TaskFinishedState.Failure;
+                // Since Execute method runs in the same process as Dev Home,
+                // we can use the previously obtained catalog package.
+                Log.Logger?.ReportInfo(Log.Component.AppManagement, $"Starting installation of package {_package.Id}");
+                var installOperator = (InstallOperator)operatorFactory.CreateInstallPackageOperator();
```
[the remark this answers is not in the document]
Comment (florelis) [reply tools/SetupFlow/DevHome.SetupFlow/Models/InstallPackageTask.cs:111]: If I'm reading this right, `CreateInstallPackageOperator()` returns an `IInstallPackageOperator` which doesn't have the `InstallPackageAsync(WinGetPackage)`, so you cast to the specific type to be able to use it, right?
Do we really need to do that? Would it be that bad to use the other version and having to search for the package again?

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        interface IApplyConfigurationResult
+            requires IOperationResult
+        {
+            // Gets the list of the configuration unit results
+            Windows.Foundation.Collections.IVector<IConfigurationUnitResult> UnitResults { get; };
+        };
+
+        interface IInstallPackageOperator
```
[the remark this answers is not in the document]
Comment (jonwis) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:61]: Maybe it's an `PackageInstallOperation` ?

Class names (things) are usually noun-phrases, like `PackageInstallOperation` and the methods on them (actions) are verb phrases like `InstallAsync`.

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            // Install package
+            Windows.Foundation.IAsyncOperation<IInstallPackageResult> InstallPackageAsync(String packageId, String catalogName);
```
Comment (jonwis) [inline tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:64]: This verb is already on an operaiton with a name containing "package" you can leave it off here - call this `InstallAsync`

File: tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            // Apply configuration file
+            Windows.Foundation.IAsyncOperation<IApplyConfigurationResult> ApplyConfigurationAsync();
+        }
+
+        // This operator exists just for testing
+        interface ITestOperator
+        {
+            // Write to standard output
+            void WriteToStdOut(String msg);
+        };
```
[the remark this answers is not in the document]
Comment (jonwis) [reply tools/SetupFlow/DevHome.SetupFlow.Contract/DevHome.SetupFlow.Contract.idl:87]: Also call this `WriteLog` - where it gets written is an implementation detail of the test operation.  And agreed that this probably shouldn't be in the list here.

**Review by jonwis — COMMENTED** [review]
Once you've addressed initial feedback, please set up time with me & @asklar  and @jamespik to go through the surface.

**Pull request closed without merging**